src/expression.rs

```diff
@@ -1,4 +1,4 @@
-use std::collections::{HashMap, HashSet};
+use std::collections::{BTreeMap, HashSet};
```

Comment:
`BTreeMap` is more efficient for our case and is compatible with `no-std`.

Reply:
Hm don't they both need std? (Sounds good either way though, if it's a bit faster.)

Reply:
Oddly it doesn't! :) There's a `no-std` implementation of `BTreeMap` that has the same API. Details are here: https://doc.rust-lang.org/stable/alloc/collections/btree_map/struct.BTreeMap.html